src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -131,6 +202,69 @@ object AST extends Positional {
     }
   }
 
+  def evalComposeCode[T](subexpr: AST)(g: Code[T] => Code[AnyRef]): CM[Code[AnyRef]] =
+    subexpr.compile().flatMap(_.mapNull((x: Code[AnyRef]) => g(x.asInstanceOf[Code[T]])))
+  def evalComposeCode[T >: Null, U >: Null](ct: Code[T])(g: Code[T] => Code[U])(implicit tti: TypeInfo[T], tct: ClassTag[T]): CM[Code[U]] =
+    ct.mapNull(g)
+  def evalComposeCodeM[T](subexpr: AST)(g: Code[T] => CM[Code[AnyRef]]): CM[Code[AnyRef]] = for (
+    (stc, c) <- CM.memoize(subexpr.compile());
```

Comment:
Memorize creates a local ref. `stc` evaluates the argument to `memoize` and stores it in a local ref. `c` loads that local ref.